File: json-schema/research/format_materialize_clock/Runner.java
// Probe: MATERIALIZE model (B) in Java via java.time. Parse each validated wire
// string into the native construct, then re-serialize via the GENERATOR-OWNED
// serializer (RFC 3339, original offset preserved with +00:00/-00:00 -> Z, T/Z
// uppercased on the parse path, fractional seconds at the value's own precision
// with trailing zeros trimmed, no fractional part when zero) -- NO TRUNCATION.
//
//   date-time -> OffsetDateTime            offset + nanosecond preserved (lossless)
//   date      -> LocalDate                 YYYY-MM-DD, lossless
//   time      -> OffsetTime / LocalTime    offset preserved when present, lossless
//
// NOTE: java.time's native toString() emits fractional seconds in 3/6/9-digit
// groups (".500", ".250") and does NOT trim trailing zeros, so we do NOT use it;
// the serializer below is generator-owned to match Go/Python/TS byte-for-byte.
//
//   java Runner.java corpus.json
import java.nio.file.*;
import java.time.*;
import java.util.regex.*;

public class Runner {
    static final String ENGINE = "java";

    static void emit(String id, String fmt, String canonical, String err) {
        System.out.println("{\"id\":\"" + id + "\",\"engine\":\"" + ENGINE
            + "\",\"format\":\"" + fmt + "\",\"canonical\":\"" + esc(canonical)
            + "\",\"err\":\"" + esc(err) + "\"}");
    }
    static String esc(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    // ".ddd" with trailing zeros trimmed, or "" when zero.
    static String fracNanos(int nanos) {
        if (nanos == 0) return "";
        String s = String.format("%09d", nanos).replaceAll("0+$", "");
        return "." + s;
    }

    // offset from a count of seconds: "Z" for zero, else "+HH:MM" / "-HH:MM".
    static String offsetStr(int secs) {
        if (secs == 0) return "Z";
        String sign = secs < 0 ? "-" : "+";
        secs = Math.abs(secs);
        return String.format("%s%02d:%02d", sign, secs / 3600, (secs % 3600) / 60);
    }

    // date-time -> OffsetDateTime (offset + nanoseconds retained; no atOffset(UTC),
    // no truncatedTo). Rejects leap :60.
    static String canonDateTime(String wire) {
        OffsetDateTime o = OffsetDateTime.parse(wire.toUpperCase());
        return String.format("%04d-%02d-%02dT%02d:%02d:%02d%s%s",
            o.getYear(), o.getMonthValue(), o.getDayOfMonth(),
            o.getHour(), o.getMinute(), o.getSecond(),
            fracNanos(o.getNano()), offsetStr(o.getOffset().getTotalSeconds()));
    }

    static String canonDate(String wire) {
        LocalDate d = LocalDate.parse(wire);
        return String.format("%04d-%02d-%02d", d.getYear(), d.getMonthValue(), d.getDayOfMonth());
    }

    // time -> OffsetTime when the wire carries an offset (RFC 3339 offset is
    // OPTIONAL), else LocalTime. Offset PRESERVED. Rejects leap :60.
    static String canonTime(String wire) {
        String w = wire.toUpperCase();
        int h, mi, s, nano, offSecs;
        boolean hasOffset;
        try {
            OffsetTime t = OffsetTime.parse(w);
            h = t.getHour(); mi = t.getMinute(); s = t.getSecond(); nano = t.getNano();
            offSecs = t.getOffset().getTotalSeconds();
            hasOffset = true;
        } catch (java.time.format.DateTimeParseException e) {
            LocalTime t = LocalTime.parse(w);
            h = t.getHour(); mi = t.getMinute(); s = t.getSecond(); nano = t.getNano();
            offSecs = 0;
            hasOffset = false;
        }
        return String.format("%02d:%02d:%02d%s%s", h, mi, s,
            fracNanos(nano), hasOffset ? offsetStr(offSecs) : "");
    }

    public static void main(String[] args) throws Exception {
        String txt = new String(Files.readAllBytes(Paths.get(args[0])));
        runArray(txt, "date-time", Runner::canonDateTime);
        runArray(txt, "date", Runner::canonDate);
        runArray(txt, "time", Runner::canonTime);
    }

    interface Fn { String apply(String s) throws Exception; }

    static void runArray(String txt, String fmt, Fn fn) {
        int key = txt.indexOf("\"" + fmt + "\"");
        int lb = txt.indexOf('[', key);
        int rb = txt.indexOf(']', lb);
        String block = txt.substring(lb, rb);
        Matcher m = Pattern.compile("\\{\\s*\"id\"\\s*:\\s*\"([^\"]*)\"\\s*,\\s*\"wire\"\\s*:\\s*\"([^\"]*)\"").matcher(block);
        while (m.find()) {
            String id = m.group(1), wire = m.group(2);
            try {
                emit(id, fmt, fn.apply(wire), "");
            } catch (Exception e) {
                emit(id, fmt, "", e.getClass().getSimpleName() + ": " + e.getMessage());
            }
        }
    }
}

File: json-schema/research/format_materialize_clock/runner.py
#!/usr/bin/env python3
"""Probe: MATERIALIZE model (B) in Python via stdlib datetime/date/time.
Parse each validated wire string into the typed construct, then re-serialize via
the GENERATOR-OWNED serializer (RFC 3339, original offset preserved with
+00:00/-00:00 -> Z, T/Z uppercased on the parse path, fractional seconds at the
value's own precision with trailing zeros trimmed, no fractional part when
zero) -- NO TRUNCATION beyond the type's genuine limit.

  date-time -> datetime (aware)      offset preserved; MICROSECOND resolution
                                     (sub-us input truncated -- the one
                                     Python-side loss).
  date      -> date                  YYYY-MM-DD, lossless.
  time      -> time (aware / naive)  offset preserved when present, lossless.

python3 runner.py corpus.json"""
import json, sys
from datetime import datetime, date, time

ENGINE = "py"


def emit(o):
    print(json.dumps(o))


def frac_micros(micro):
    """'.ddd' with trailing zeros trimmed, or '' when zero."""
    if micro == 0:
        return ""
    return "." + f"{micro:06d}".rstrip("0")


def offset_str(td):
    """Offset from a timedelta: 'Z' for zero, '' when None, else +/-HH:MM."""
    if td is None:
        return ""
    secs = int(td.total_seconds())
    if secs == 0:
        return "Z"
    sign = "+" if secs > 0 else "-"
    secs = abs(secs)
    return f"{sign}{secs // 3600:02d}:{(secs % 3600) // 60:02d}"


def canon_datetime(wire):
    # Parse path uppercases case-insensitive t/z (pinned grammar accepts
    # lowercase; fromisoformat rejects it). Safe: no other letters. Rejects :60.
    dt = datetime.fromisoformat(wire.upper())
    if dt.tzinfo is None:
        raise ValueError("missing offset (naive) -- date-time requires an offset")
    # offset PRESERVED (no astimezone(utc)); microsecond resolution is the
    # native limit -- any finer input was already truncated by fromisoformat.
    return (f"{dt.year:04d}-{dt.month:02d}-{dt.day:02d}"
            f"T{dt.hour:02d}:{dt.minute:02d}:{dt.second:02d}"
            f"{frac_micros(dt.microsecond)}{offset_str(dt.utcoffset())}")


def canon_date(wire):
    d = date.fromisoformat(wire)
    return f"{d.year:04d}-{d.month:02d}-{d.day:02d}"


def canon_time(wire):
    t = time.fromisoformat(wire.upper())  # aware when offset present, else naive
    return (f"{t.hour:02d}:{t.minute:02d}:{t.second:02d}"
            f"{frac_micros(t.microsecond)}{offset_str(t.utcoffset())}")


def run(rows, fmt, fn):
    for r in rows:
        try:
            emit({"id": r["id"], "engine": ENGINE, "format": fmt,
                  "canonical": fn(r["wire"]), "err": ""})
        except Exception as e:
            emit({"id": r["id"], "engine": ENGINE, "format": fmt,
                  "canonical": "", "err": f"{type(e).__name__}: {e}"})


def main():
    c = json.load(open(sys.argv[1]))
    run(c["date-time"], "date-time", canon_datetime)
    run(c["date"], "date", canon_date)
    run(c["time"], "time", canon_time)


if __name__ == "__main__":
    main()
